docs/docsite/rst/dev_guide/developing_module_utilities.rst

```diff
@@ -4,75 +4,142 @@
 Appendix: Module Utilities
 **************************
 
-Ansible provides a number of module utilities that provide helper functions that you can use when developing your own modules. The ``basic.py`` module utility provides the main entry point for accessing the Ansible library, and all Python Ansible modules must, at minimum, import ``AnsibleModule``::
+A mechanism exists to ship utility modules imported by a :ref:`New-Style
+Module <flow_new_style_modules>` alongside for execution on the target. When a
+new-style module executes, utility modules may be imported via the
+``ansible.module_utils`` Python package namespace. The module entry point is
+imported from this namespace, and must always be present:
+
+.. code-block:: python
 
   from ansible.module_utils.basic import AnsibleModule
 
+The ``ansible.module_utils`` namespace is not a plain Python package: it is
+constructed dynamically for each task invocation, by extracting imports and
+resolving those matching the namespace against a search path derived from the
+active configuration.
+
+Module utilities tree structure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A Python module utility must reuse same namespace/directories than related
+modules. For example, EMC modules belong to the ``storage/emc`` namespace,
+meaning they are located in ``lib/ansible/modules/storage/emc/``. Then common
+code to EMC modules must be located in the ``lib/ansible/modules/storage/emc/``
+directory.
+
+Search Path
+~~~~~~~~~~~
```

Comment:
This section looks like it doesn't belong here.  It's relevant for all plugins, not just module_utils